File: scripts/collect_naver_board.ts
/**
 * 네이버 종목토론 수집 — 한세 4종목.
 *
 * 배경:
 *   기존 /api/cron/naver-board가 Vercel Hobby plan 60s 한계에 걸려 504.
 *   GHA runner에서 직접 fetchNaverBoardPosts() 호출 후 Supabase upsert로 전환
 *   (다른 collect_*.py 워크플로와 동일한 패턴, Vercel timeout 무관).
 *
 * 🔴 종료코드로 「조용한 정지」를 막는다(2026-09-12 개편 때 나흘을 잃고 신설):
 *   0 = 정상 · 1 = 환경·DB 문제 · 3 = 수집 경로가 깨졌다(구조 변경 의심)
 *   exit 3 은 종목이 하나라도 실패했거나, **네 종목 모두 네이버가 글을 0건 준** 경우다.
 *   글이 0건인 것과 API 가 빈 배열을 주는 것을 가르려고 rawCount 를 함께 본다.
 *
 * 실행: npx tsx scripts/collect_naver_board.ts
 * 환경: SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY
 */
import { createClient } from '@supabase/supabase-js';
import { fetchNaverBoardPosts } from '../lib/naver/board';

const HANSAE_TICKERS = ['016450', '105630', '069640', '053280'];
/** 수집 경로가 깨진 것으로 보고 실패시키는 종료코드. */
const EXIT_BROKEN = 3;

async function main(): Promise<void> {
  const url = process.env.SUPABASE_URL ?? process.env.NEXT_PUBLIC_SUPABASE_URL;
  const key = process.env.SUPABASE_SERVICE_ROLE_KEY;
  if (!url || !key) {
    console.error('SUPABASE_URL/SUPABASE_SERVICE_ROLE_KEY 미설정');
    process.exit(1);
  }
  const sb = createClient(url, key, { auth: { persistSession: false } });

  const { data: companies, error } = await sb
    .from('companies')
    .select('id,ticker')
    .in('ticker', HANSAE_TICKERS);
  if (error || !companies) {
    console.error('companies 조회 실패:', error?.message);
    process.exit(1);
  }

  const summary: { ticker: string; inserted: number; total: number; raw: number }[] = [];
  const failed: string[] = [];
  for (const c of companies as { id: string; ticker: string }[]) {
    try {
      // 본문은 개편 뒤 목록 응답에 함께 오지만(추가 요청 없음) 감성 분석은 제목만
      // 쓰므로 종전대로 담지 않는다.
      const { posts, rawCount } = await fetchNaverBoardPosts(c.ticker, 7, 10, false);
      if (posts.length === 0) {
        summary.push({ ticker: c.ticker, inserted: 0, total: 0, raw: rawCount });
        continue;
      }
      const rows = posts.map((p) => ({
        company_id: c.id,
        post_id: p.postId,
        posted_at: p.postedAt.toISOString(),
        title: p.title,
        body: p.body,
        views: p.views,
        likes: p.likes,
        dislikes: p.dislikes,
        fetched_at: new Date().toISOString(),
      }));
      const { count, error: upErr } = await sb
        .from('naver_board_posts')
        .upsert(rows, { onConflict: 'company_id,post_id', count: 'exact' });
      if (upErr) {
        console.error(`${c.ticker} upsert 실패:`, upErr.message);
        failed.push(c.ticker);
        summary.push({ ticker: c.ticker, inserted: 0, total: posts.length, raw: rawCount });
        continue;
      }
      summary.push({
        ticker: c.ticker,
        inserted: count ?? rows.length,
        total: posts.length,
        raw: rawCount,
      });
    } catch (err) {
      console.error(`${c.ticker} 수집 실패:`, err);
      failed.push(c.ticker);
      summary.push({ ticker: c.ticker, inserted: 0, total: 0, raw: 0 });
    }
  }

  console.log('네이버 종목토론 수집 완료:', JSON.stringify(summary));

  const revUrl = process.env.NEXT_REVALIDATE_URL;
  const revSecret = process.env.NEXT_REVALIDATE_SECRET;
  if (revUrl && revSecret) {
    try {
      const r = await fetch(revUrl, {
        method: 'POST',
        headers: { 'x-revalidate-secret': revSecret, 'Content-Type': 'application/json' },
        body: JSON.stringify({ tags: ['naver_board_posts'] }),
      });
      console.log('revalidate:', r.status);
    } catch (err) {
      console.warn('revalidate 실패:', err);
    }
  }

  // 🔴 여기서 조용히 0 으로 끝내지 않는다 — 옛 수집기가 그래서 나흘을 잃었다.
  if (failed.length > 0) {
    console.error(`::error::종목토론 수집 실패 ${failed.join(', ')} — 수집 경로 확인 필요`);
    process.exit(EXIT_BROKEN);
  }
  if (summary.every((s) => s.raw === 0)) {
    console.error('::error::네 종목 모두 네이버가 글을 0건 반환 — 구조가 또 바뀌었을 수 있습니다');
    process.exit(EXIT_BROKEN);
  }
}

main().catch((err) => {
  console.error('치명적 오류:', err);
  process.exit(1);
});

File: scripts/collect_naver_board.py
#!/usr/bin/env python3
"""
한세 4종목(016450/105630/069640/053280)의 네이버 금융 종목토론실 글을 스크래핑해
naver_board_posts 테이블에 upsert한다.

URL 구조:
  https://finance.naver.com/item/board.naver?code={ticker}&page={n}

각 글의 nid(post_id)·작성일시·제목·조회수·공감·비공감만 목록 페이지에서 수집.
본문은 비용 대비 가치가 낮아 생략 (body=NULL).

수집 모드 (--mode):
- full         : 1~MAX_PAGES 페이지 (기본 5페이지)
- incremental  : 마지막 수집 post_id를 만나면 중단 (기본)
"""
import argparse
import re
import sys
import time
from datetime import datetime, timezone, timedelta
from pathlib import Path

import requests
from bs4 import BeautifulSoup
from dotenv import load_dotenv
from loguru import logger

load_dotenv(Path(__file__).parent / '.env')
load_dotenv(Path(__file__).parent.parent / '.env.local')

from lib.db import get_client, upsert_rows

HANSAE_TICKERS = ['016450', '105630', '069640', '053280']
HEADERS = {
  'User-Agent': (
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 '
    '(KHTML, like Gecko) Chrome/119.0.0.0 Safari/537.36'
  ),
  'Referer': 'https://finance.naver.com/',
}
KST = timezone(timedelta(hours=9))
MAX_PAGES_DEFAULT = 5
PAGE_SLEEP = 0.5  # 페이지 간 sleep (부담 최소화)


def _load_hansae_companies() -> list[dict]:
  client = get_client()
  res = (
    client.table('companies')
    .select('id,ticker,name_kr')
    .in_('ticker', HANSAE_TICKERS)
    .execute()
  )
  return res.data


def _load_known_post_ids(company_id: str, since_days: int = 7) -> set[str]:
  """최근 N일 안의 post_id 집합. incremental 중단 판단용."""
  client = get_client()
  since = (datetime.now(KST) - timedelta(days=since_days)).isoformat()
  res = (
    client.table('naver_board_posts')
    .select('post_id')
    .eq('company_id', company_id)
    .gte('posted_at', since)
    .execute()
  )
  return {r['post_id'] for r in (res.data or [])}


def _parse_int(s: str) -> int | None:
  s = s.strip().replace(',', '')
  if not s or s == '-':
    return None
  try:
    return int(s)
  except ValueError:
    return None


def _parse_posted_at(s: str) -> str | None:
  """네이버 표기 → ISO KST. '2026.05.17 10:30' 또는 '2026.05.17' 형식."""
  s = s.strip()
  for fmt in ('%Y.%m.%d %H:%M', '%Y.%m.%d'):
    try:
      dt = datetime.strptime(s, fmt).replace(tzinfo=KST)
      return dt.isoformat()
    except ValueError:
      pass
  return None


_NID_RE = re.compile(r'nid=(\d+)')


def _fetch_page(ticker: str, page: int) -> list[dict]:
  """종목토론실 한 페이지의 글 목록을 dict 배열로 반환."""
  url = f'https://finance.naver.com/item/board.naver?code={ticker}&page={page}'
  r = requests.get(url, headers=HEADERS, timeout=15)
  # 네이버 금융은 EUC-KR
  r.encoding = r.apparent_encoding or 'euc-kr'
  soup = BeautifulSoup(r.text, 'html.parser')

  posts: list[dict] = []
  rows = soup.select('table.type2 tr[onmouseover]')
  if not rows:
    # 일부 환경은 onmouseover 속성이 없을 수 있어 대안 — td 첫 번째에 날짜 패턴 포함
    rows = [tr for tr in soup.select('table.type2 tr') if tr.select_one('td span.tah')]

  for tr in rows:
    tds = tr.find_all('td')
    if len(tds) < 6:
      continue
    a_tag = tds[1].find('a', href=True)
    if not a_tag:
      continue
    m = _NID_RE.search(a_tag['href'])
    if not m:
      continue
    post_id = m.group(1)
    title = a_tag.get_text(strip=True)
    posted = tds[0].get_text(strip=True)
    views = tds[3].get_text(strip=True)
    likes = tds[4].get_text(strip=True)
    dislikes = tds[5].get_text(strip=True)
    posts.append({
      'post_id': post_id,
      'title': title,
      'posted_at': _parse_posted_at(posted),
      'views': _parse_int(views),
      'likes': _parse_int(likes),
      'dislikes': _parse_int(dislikes),
    })
  return posts


def collectNaverBoard(mode: str = 'incremental', max_pages: int = MAX_PAGES_DEFAULT) -> None:
  companies = _load_hansae_companies()
  logger.info(f'종목토론 수집 모드: {mode}, max_pages={max_pages}, 종목 {len(companies)}개')

  total_rows = 0
  for company in companies:
    ticker = company['ticker']
    known = _load_known_post_ids(company['id']) if mode == 'incremental' else set()
    new_rows: list[dict] = []
    now_iso = datetime.now(KST).isoformat()

    for page in range(1, max_pages + 1):
      try:
        posts = _fetch_page(ticker, page)
      except Exception as e:
        logger.warning(f'{ticker} page={page} fetch 실패: {e}')
        continue
      if not posts:
        logger.debug(f'{ticker} page={page}: 빈 목록 — 중단')
        break

      page_new = 0
      for p in posts:
        if not p.get('post_id') or not p.get('posted_at'):
          continue
        if mode == 'incremental' and p['post_id'] in known:
          continue
        new_rows.append({
          'company_id': company['id'],
          'post_id': p['post_id'],
          'posted_at': p['posted_at'],
          'title': p['title'],
          'body': None,
          'views': p['views'],
          'likes': p['likes'],
          'dislikes': p['dislikes'],
          'fetched_at': now_iso,
        })
        page_new += 1

      logger.debug(f'{ticker} page={page}: {len(posts)}건 중 신규 {page_new}건')
      # incremental 모드: 페이지 전체가 known이면 더 안 봐도 됨
      if mode == 'incremental' and page_new == 0:
        break
      time.sleep(PAGE_SLEEP)

    if new_rows:
      upsert_rows('naver_board_posts', new_rows, 'company_id,post_id')
      total_rows += len(new_rows)
      logger.info(f"{ticker} ({company['name_kr']}): {len(new_rows)}건 upsert")
    else:
      logger.info(f"{ticker} ({company['name_kr']}): 신규 없음")

  logger.info(f'종목토론 수집 완료 — 총 {total_rows}건')


if __name__ == '__main__':
  parser = argparse.ArgumentParser(description='한세 3종목 네이버 종목토론 수집')
  parser.add_argument(
    '--mode', choices=['full', 'incremental'], default='incremental',
    help='full: 1~max_pages 모두, incremental: 기존 post_id 만나면 중단 (기본값)'
  )
  parser.add_argument('--max-pages', type=int, default=MAX_PAGES_DEFAULT)
  args = parser.parse_args()
  try:
    collectNaverBoard(mode=args.mode, max_pages=args.max_pages)
  except Exception as e:
    import traceback
    logger.error(f'종목토론 수집 실패: {e}\n{traceback.format_exc()}')
    sys.exit(1)
